### ISSUE TITLE
Found with fuzzer: fearedmon_id problem in savestateinlock

### ISSUE THREAD

**Issue by elunna** [elunna/hackem#37]
```
=================================================================
==4339==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000211150 at pc 0x5555560b09f5 bp 0x7fffffffe0f0 sp 0x7fffffffe0e0
READ of size 4 at 0x611000211150 thread T0
    #0 0x5555560b09f4 in savestateinlock /home/lunatunez/hackem/src/save.c:465
    #1 0x555555911d67 in save_currentstate /home/lunatunez/hackem/src/do.c:1612
    #2 0x5555558350c7 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:923
    #3 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #4 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #5 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #6 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308
    #7 0x55555575dacd in _start (/home/lunatunez/games/hackemdir/hackem+0x209acd)

0x611000211150 is located 16 bytes inside of 216-byte region [0x611000211140,0x611000211218)
freed by thread T0 here:
    #0 0x7ffff769940f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x555555dafb58 in dealloc_monst /home/lunatunez/hackem/src/mon.c:3169
    #2 0x5555560b4369 in savemonchn /home/lunatunez/hackem/src/save.c:1263
    #3 0x5555560b1151 in savelev /home/lunatunez/hackem/src/save.c:599
    #4 0x555555834e98 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:893
    #5 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #6 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #7 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #8 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

previously allocated by thread T0 here:
    #0 0x7ffff7699808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x55555576a2e4 in alloc /home/lunatunez/hackem/src/alloc.c:46
    #2 0x555555b57a5a in makemon /home/lunatunez/hackem/src/makemon.c:3124
    #3 0x5555560850db in create_particular_creation /home/lunatunez/hackem/src/read.c:3706
    #4 0x55555608a8c2 in create_particular /home/lunatunez/hackem/src/read.c:3831
    #5 0x555555835955 in wiz_genesis /home/lunatunez/hackem/src/cmd.c:980
    #6 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #7 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #8 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #9 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-use-after-free /home/lunatunez/hackem/src/save.c:465 in savestateinlock
Shadow bytes around the buggy address:
  0x0c228003a1d0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c228003a1e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228003a1f0: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c228003a200: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228003a210: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
=>0x0c228003a220: fa fa fa fa fa fa fa fa fd fd[fd]fd fd fd fd fd
  0x0c228003a230: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228003a240: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c228003a250: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c228003a260: 00 00 00 00 00 00 00 00 00 00 00 fa fa fa fa fa
  0x0c228003a270: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==4339==ABORTING
```

**Comment by elunna** [comment]
Another similar capture:
```
=================================================================
==4355==ERROR: AddressSanitizer: heap-use-after-free on address 0x6110004d3a50 at pc 0x5555560b09f5 bp 0x7fffffffe0f0 sp 0x7fffffffe0e0
READ of size 4 at 0x6110004d3a50 thread T0
    #0 0x5555560b09f4 in savestateinlock /home/lunatunez/hackem/src/save.c:465
    #1 0x555555911d67 in save_currentstate /home/lunatunez/hackem/src/do.c:1612
    #2 0x5555558350c7 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:923
    #3 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #4 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #5 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #6 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308
    #7 0x55555575dacd in _start (/home/lunatunez/games/hackemdir/hackem+0x209acd)

0x6110004d3a50 is located 16 bytes inside of 216-byte region [0x6110004d3a40,0x6110004d3b18)
freed by thread T0 here:
    #0 0x7ffff769940f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x555555dafb58 in dealloc_monst /home/lunatunez/hackem/src/mon.c:3169
    #2 0x5555560b4369 in savemonchn /home/lunatunez/hackem/src/save.c:1263
    #3 0x5555560b1151 in savelev /home/lunatunez/hackem/src/save.c:599
    #4 0x555555834e98 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:893
    #5 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #6 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #7 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #8 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

previously allocated by thread T0 here:
    #0 0x7ffff7699808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x55555576a2e4 in alloc /home/lunatunez/hackem/src/alloc.c:46
    #2 0x555555b57a5a in makemon /home/lunatunez/hackem/src/makemon.c:3124
    #3 0x55555632a43b in nasty /home/lunatunez/hackem/src/wizard.c:828
    #4 0x555555b8400e in cast_wizard_spell /home/lunatunez/hackem/src/mcastu.c:781
    #5 0x555555b7b37b in castmu /home/lunatunez/hackem/src/mcastu.c:494
    #6 0x555555e3fb03 in dochug /home/lunatunez/hackem/src/monmove.c:903
    #7 0x555555e2bae2 in dochugw /home/lunatunez/hackem/src/monmove.c:122
    #8 0x555555d99ceb in movemon /home/lunatunez/hackem/src/mon.c:1517
    #9 0x55555576023a in moveloop /home/lunatunez/hackem/src/allmain.c:244
    #10 0x5555563ca091 in main ../sys/unix/unixmain.c:353
    #11 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-use-after-free /home/lunatunez/hackem/src/save.c:465 in savestateinlock
Shadow bytes around the buggy address:
  0x0c22800926f0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c2280092700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280092710: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2280092720: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280092730: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
=>0x0c2280092740: fa fa fa fa fa fa fa fa fd fd[fd]fd fd fd fd fd
  0x0c2280092750: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280092760: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2280092770: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280092780: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c2280092790: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==4355==ABORTING
```

**Comment by elunna** [comment]
5970ea059] Fix for issue #37: Fixed missing free for fearedmon_id while saving.

**Comment by elunna** [comment]
Popped up again after previous fix.
```
=================================================================
==28522==ERROR: AddressSanitizer: heap-use-after-free on address 0x6110002dea10 at pc 0x5555560b0a24 bp 0x7fffffffe0f0 sp 0x7fffffffe0e0
READ of size 4 at 0x6110002dea10 thread T0
    #0 0x5555560b0a23 in savestateinlock /home/lunatunez/hackem/src/save.c:466
    #1 0x555555911d67 in save_currentstate /home/lunatunez/hackem/src/do.c:1612
    #2 0x5555558350c7 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:923
    #3 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #4 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #5 0x5555563ca0c0 in main ../sys/unix/unixmain.c:353
    #6 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308
    #7 0x55555575dacd in _start (/home/lunatunez/games/hackemdir/hackem+0x209acd)

0x6110002dea10 is located 16 bytes inside of 216-byte region [0x6110002dea00,0x6110002dead8)
freed by thread T0 here:
    #0 0x7ffff769940f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x555555dafb58 in dealloc_monst /home/lunatunez/hackem/src/mon.c:3169
    #2 0x5555560b4398 in savemonchn /home/lunatunez/hackem/src/save.c:1265
    #3 0x5555560b1180 in savelev /home/lunatunez/hackem/src/save.c:600
    #4 0x555555834e98 in wiz_makemap /home/lunatunez/hackem/src/cmd.c:893
    #5 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #6 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #7 0x5555563ca0c0 in main ../sys/unix/unixmain.c:353
    #8 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

previously allocated by thread T0 here:
    #0 0x7ffff7699808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x55555576a2e4 in alloc /home/lunatunez/hackem/src/alloc.c:46
    #2 0x555555b57a5a in makemon /home/lunatunez/hackem/src/makemon.c:3124
    #3 0x5555560850db in create_particular_creation /home/lunatunez/hackem/src/read.c:3706
    #4 0x55555608a8c2 in create_particular /home/lunatunez/hackem/src/read.c:3831
    #5 0x555555835955 in wiz_genesis /home/lunatunez/hackem/src/cmd.c:980
    #6 0x555555866bd0 in rhack /home/lunatunez/hackem/src/cmd.c:5529
    #7 0x555555766f6b in moveloop /home/lunatunez/hackem/src/allmain.c:821
    #8 0x5555563ca0c0 in main ../sys/unix/unixmain.c:353
    #9 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-use-after-free /home/lunatunez/hackem/src/save.c:466 in savestateinlock
Shadow bytes around the buggy address:
  0x0c2280053cf0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280053d00: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c2280053d10: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c2280053d20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280053d30: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c2280053d40: fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280053d50: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c2280053d60: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c2280053d70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280053d80: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2280053d90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==28522==ABORTING
```

**Comment by elunna** [comment]
Cannot reproduce by:
Being afflicted with Afraid by a banshee and #wizmap
or
Breaking a wand of fear and #wizmap (I suspected this since it might not set the monster to be afraid of, but nope)

**Comment by elunna** [comment]
Commit 7b3f2a0a4 Fix for issue #37. The problem with fearedmon seemed isloated to when makewizmap was called. I cleared the fear status and feared monster whenever a new map is generated.
